### PR TITLE
add Resol VBus support

### LINKED ISSUE #1949: Resol Deltasol support
**Describe the problem you have/What new integration you would like**

Support for receive data over Rx from Resol Deltasol solar controller. 

**Please describe your use case for this integration and alternatives you've tried:**

I need data to home assistant
Alterane of  Tasmota smart meter and his script for Resol

**Additional context**

### PR DESCRIPTION
**Pull request by ssieb**
# What does this implement/fix?

This component adds support for several Resol VBus devices and a custom option to allow arbitrary parsing of not yet supported devices.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/feature-requests/issues/1949

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2397

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by nagyrobi** [comment]
As the conversation in https://github.com/esphome/feature-requests/issues/1949 shows, this component works flawlessly.
Tested with today's `dev`.

File: esphome/components/vbus/sensor/vbus_sensor.cpp [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    ESP_LOGCONFIG(TAG, "  Command: ANY");
+  } else {
+    ESP_LOGCONFIG(TAG, "  Command: 0x%04x", this->command_);
+  }
+}
+
+void VBusCustomSensor::handle_message(std::vector<uint8_t> &message) {
+  if (this->message_handler_.has_value())
+    (*this->message_handler_)(message);
+}
```
Comment (jesserockz) [inline esphome/components/vbus/sensor/vbus_sensor.cpp]: Would it not be better to take the return value from this lambda and then publish it as the state?

Comment (ssieb) [reply esphome/components/vbus/sensor/vbus_sensor.cpp]: `message` is not a single value.  It's a packet with all the values.  That message handler parses the message and publishes to all the sensors that are configured.

Comment (ssieb) [reply esphome/components/vbus/sensor/vbus_sensor.cpp]: I was misreading which method that was.  I will change how this works.

**Comment by ssieb** [comment]
The test compile failure is from a different component.  Compiler change?

File: tests/test5.yaml [hunk 2]
```diff
@@ -478,6 +497,27 @@ sensor:
...
+
+  - platform: vbus
+    model: custom
+    command: 0x100
+    source: 0x1234
+    dest: 0x10
+    sensors:
+      - id: vcustom
+        name: VBus Custom Sensor
+        lambda: return x[0] / 10.0
```
Comment (jesserockz) [inline tests/test5.yaml]: CI error due to missing semicolon
```suggestion
        lambda: return x[0] / 10.0;
```

**Comment by Gloomyeye** [comment]
I didn't test the coding of this PR, but used the coding of the ssieb's external component and the PR documentation (https://deploy-preview-2397--esphome.netlify.app/components/vbus.html) to build and run successfully an esp32 with a DeltaSol CS Plus device.

I am looking forward to see this directly included in the esphome version. :-)

This is my config:
```# Resol VBUS
external_components:
  - source:
      type: git
      url: https://github.com/ssieb/custom_components
      ref: vbus
    components: [ vbus ]
    refresh: 1min

uart:
  id: resol
  rx_pin: GPIO20
  baud_rate: 9600

vbus:
  uart_id: resol

**Comment by SeByDocKy** [comment]
> external_components:
>   - source:
>       type: git
>       url: https://github.com/ssieb/custom_components
>       ref: vbus
>     components: [ vbus ]
>     refresh: 1min

Thx it helped me a lot ... coz direct PR inclusion was not working

**Review by jesserockz — APPROVED** [review]

**Pull request merged**